### PR TITLE
Add submission count property to interactive window telemetry

### PR DESCRIPTION
**Pull request by genlu**
Add a property for telemetry event of interactive window.

The affected scenario is when VS closed or an instance of interactive window crashed. In this case, a  `VS.IDE.VBCS.CSharp.Interactive.Window.InteractiveWindow` property is set to `Close` 
and a `VS.IDE.VBCS.CSharp.Interactive.Window.LanguageBufferCount` property is set to the total number of language buffer created in one interactive window session (i.e. the lifetime of one instance), this property counts both code submissions and command submissions.

Also removed event logging for interactive window opening.

### CODE REVIEW

**Comment by genlu** [comment]
@ManishJayaswal @amcasey @cston @KevinH-MS @tmat 

FYI @heejaechang @kuhlenh

File: src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs [hunk 1]
```diff
@@ -64,9 +66,23 @@ protected override string Title
                 CommonVsUtils.GetWorkingDirectory());
         }
 
-        protected override void LogSession(string key, string value)
+        protected override void LogSession(string key, string value, params object[] objects)
         {
-            Logger.Log(FunctionId.CSharp_Interactive_Window, KeyValueLogMessage.Create(m => m.Add(key, value)));
+            Action<Dictionary<string, object>> propertySetter;
+            if (value.Equals(LanguageServices.Interactive.LogMessage.Close))
```
Comment (heejaechang) [inline src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs]: value never null I guess? why not using == ?

File: src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs [hunk 2]
```diff
@@ -64,9 +66,23 @@ protected override string Title
...
+            {
+                Debug.Assert(objects.Length > 0);
+                propertySetter = m =>
+                                    {
+                                        m.Add(key, value);
+                                        m.Add(LanguageServices.Interactive.LogMessage.SubmissionCount, objects[0]);
```
Comment (heejaechang) [inline src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs]: I would rather create LogCloseSession and get int submissionCount as parameter.

File: src/VisualStudio/InteractiveServices/Interactive/VsInteractiveWindowProvider.cs [hunk 3]
```diff
@@ -58,7 +58,7 @@ internal abstract class VsInteractiveWindowProvider
         protected abstract Guid LanguageServiceGuid { get; }
         protected abstract Guid Id { get; }
         protected abstract string Title { get; }
-        protected abstract void LogSession(string key, string value);
+        protected abstract void LogSession(string key, string value, params object[] objects);
```
Comment (heejaechang) [inline src/VisualStudio/InteractiveServices/Interactive/VsInteractiveWindowProvider.cs]: actual implementation of this method seems static. and doesnt actually require behavior different between derived type. how about just creating static type InteractiveLogger.Log( ... )

File: src/VisualStudio/VisualBasic/Repl/VisualBasicVsInteractiveWindowProvider.vb [hunk 4]
```diff
@@ -65,8 +65,18 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Interactive
                 CommonVsUtils.GetWorkingDirectory())
         End Function
 
-        Protected Overrides Sub LogSession(key As String, value As String)
-            Logger.Log(FunctionId.VisualBasic_Interactive_Window, KeyValueLogMessage.Create(Sub(m) m.Add(key, value)))
+        Protected Overrides Sub LogSession(key As String, value As String, ParamArray objects() As Object)
```
Comment (heejaechang) [inline src/VisualStudio/VisualBasic/Repl/VisualBasicVsInteractiveWindowProvider.vb]: this seems a duplicated implementation.

**Comment by heejaechang** [comment]
:+1:

File: src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs (hunk 1 above)
Comment (genlu) [inline src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs]: Good point, and we are comparing constants here. Fixed, thanks!

File: src/VisualStudio/InteractiveServices/Interactive/VsInteractiveWindowProvider.cs (hunk 3 above)
Comment (genlu) [inline src/VisualStudio/InteractiveServices/Interactive/VsInteractiveWindowProvider.cs]: I think the difference between derived type is the `FunctionId` passed to `Logger.Log`.

File: src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs (hunk 2 above)
Comment (genlu) [inline src/VisualStudio/CSharp/Repl/CSharpVsInteractiveWindowProvider.cs]: Sure, fixed.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 5]
```diff
@@ -832,6 +832,7 @@ private static int IndexOfNonWhiteSpaceCharacter(ITextSnapshotLine line)
                         {
                             FinishExecute(executionResult.IsSuccessful);
                         }
+                        _window.SubmissionCounter++;
```
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Will this count reset or clear or history command if the user used the UI to execute those?

**Comment by ManishJayaswal** [comment]
LGTM except for the comment above.

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs (hunk 5 above)
Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Hmm... Just realized you can actually do that using UI, will verify/fix. Thanks!

Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: It doesn't count those commands executed through UI. It also doesn't count any commands executed through keyboard shortcuts.

Will fix for `Reset` and `Clear Screen` UI command.

**Comment by genlu** [comment]
@dotnet-bot Retest this, please

File: src/VisualStudio/VisualBasic/Repl/VisualBasicVsInteractiveWindowProvider.vb [hunk 6]
```diff
@@ -68,6 +68,14 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Interactive
         Protected Overrides Sub LogSession(key As String, value As String)
             Logger.Log(FunctionId.VisualBasic_Interactive_Window, KeyValueLogMessage.Create(Sub(m) m.Add(key, value)))
         End Sub
+
+        Protected Overrides Sub LogCloseSession(languageBufferCount As Integer)
+            Logger.Log(FunctionId.VisualBasic_Interactive_Window,
```
Comment (amcasey) [inline src/VisualStudio/VisualBasic/Repl/VisualBasicVsInteractiveWindowProvider.vb]: Is this the same as VB, other than the ```FunctionId```?  If so, should we just have an abstract ```FunctionId``` property?

**Comment by amcasey** [comment]
:+1:

File: src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs [hunk 7]
```diff
@@ -1386,6 +1386,8 @@ private void AddLanguageBuffer()
                     bufferAdded(_window, new SubmissionBufferAddedEventArgs(buffer));
                 }
 
+                _window.LanguageBufferCounter++;
+
```
Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Are we counting everytime AddLanguageBuffer method is called? Shouldn't this counter be updated in that method instead?

Comment (genlu) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: @ManishJayaswal What method?

Comment (ManishJayaswal) [inline src/InteractiveWindow/Editor/InteractiveWindow.UIThreadOnly.cs]: Per offline conversation, this is exactly in the method that I mentioned above so it is good.  I was misguided when the CR view showed this under SubmitAsync.

**Pull request merged**